### ISSUE TITLE
Steam resetting permissions when switching between Windows and Arch Linux

### ISSUE THREAD

**Issue by markmann447** [maharmstone/btrfs#655]
Hey, I don't know if this is a winbtrfs thing or a Valve thing, but basically when I play a game on an extra drive formatted using btrfs. When I log back into Linux, I have to take ownership of each drive with steam games on it, and this happens every time. Is there anything you can do on your end to stop the permissions from resetting when logging back into Linux? The reason I'm opening up this issue is because there's a potential risk of breaking my system when taking permissions on my Arch Linux drive. Obviously the way to prevent this from happening is to not format the OS drive, but in my personal opinion winbtrfs is a MUST HAVE tool for people who dual boot, and having this tool would make the linux adoption much higher for people who still need Windows on a seperate drive when they play games with intrusive anti-cheat or need to use programs that just don't work under WINE.

Thank you regardless for this wonderful tool.

**Comment by rautamiekka** [comment]
What do you mean by take ownership/permissions ?

**Comment by markmann447** [comment]
![missing file privilges](https://github.com/maharmstone/btrfs/assets/94986033/3ff5186f-c2dc-4300-a1aa-b6af743ada2f)
So whenever I would switch to my Windows partition to play a game on Windows, it would run fine and update. Although when I switch to Linux, I get this  "Missing file privileges" error. The way to solve this issue is to basically go into gnome disk utility and "Take Ownership" of each individual drive. This basically happens because Windows on Steam is somehow removing or resetting the file permissions for games on btrfs formatted drives. I was wondering if there was anything you can do on your end to solve this issue.

**Comment by 42Willow** [comment]
This is probably related #421

**Comment by weskerty** [comment]
> 
> > Have a look at the Mappings section of the README, which tells you how to map your Windows SID to your Linux uid.

would this be?

> > SubvolId (QWORD): the ID of the subvolume that we will attempt to mount as the root. If it doesn't exist, this parameter will be silently ignored. The subvolume ID can be found on the inode property sheet; it's in hex there, as opposed to decimal on the Linux tools. The default is whatever has been set via btrfs subvolume set-default; or, failing that, subvolume 5. The equivalent parameter on Linux is called subvolid.

**Comment by weskerty** [comment]
then it would be to create a DWORD entry in the registry
HKLM\SYSTEM\CurrentControlSet\services\btrfs\Mappings with the id of your user in Windows plus the value of the user in Linux?
![image](https://github.com/maharmstone/btrfs/assets/82781997/de6b324b-fb19-477b-95a8-c153de4a7e95)
![image](https://github.com/maharmstone/btrfs/assets/82781997/2964e616-99a1-4092-b335-d59c6badaeb7)

**Comment by 42Willow** [comment]
I'm also experiencing the same issue, I just haven't been bothered trying to fix it with the solution I referenced...

This is what I'm going to try now:

**SID (windows):** `wmic useraccount get name,sid`
**UID (linux):** `$ id`
**GID (linux):** `$ id`

Arch Linux has a default group ID of `1000`, but the default is `100` (according to the README)
> The default entry maps Windows' Users group to gid 100, which is usually "users" on Linux.

### Windows
User mappings: `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\btrfs\Mappings`
Group mappings: `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\btrfs\GroupMappings`

**repeat for both user mappings and group mappings**
1. Create a new DWORD (32-bit) for your `SID`:
2. Name the DWORD with your Windows `SID`
3. Set the value to your Linux `UID` or `GID`
4. Repeat one more time for `UID` or `GID`

I'll get back to you to let you know if it works

**Comment by 42Willow** [comment]
@weskerty I'm having trouble with the permissions, where did you find that menu?

**Comment by weskerty** [comment]
> @weskerty I'm having trouble with the permissions, where did you find that menu?

file properties, Btrfs tab

**Comment by weskerty** [comment]
Manage to adjust to change the user.
In \btrfs\GroupMappings there was a DWORD S-1-5-32-545
There I set to Users (which in my file is 984)
![Screenshot_2](https://github.com/maharmstone/btrfs/assets/82781997/b49f8194-6989-4829-82fb-245f5b208972)

But I can't Change the Owner, I added all the users, I also cloned the GroupMappings ID and still nothing changes.
The new folder and file created from Windows will have the ID 65534 which appears as Nobody in ArchLinux.
![image](https://github.com/maharmstone/btrfs/assets/82781997/fb076e29-0aec-4815-af2c-8e0f4b61c4a3)
![image](https://github.com/maharmstone/btrfs/assets/82781997/d363ba5c-1234-4fa2-85cc-2d4d8e8e0ee9)


I don't know if I should open another issue about this.

**Comment by youbloub** [comment]
same error for me.

installing game under linux,
launch it: working
going on windows( with WinBtrfs v1.9) : working.
going back to linux: get error : "an error occured while launching this game: missing file privileges"

tested on 2 computers and get same error.
tried following git winbtrfs wiki : the mapping and fast boot added in regedit  does not help.
making a chown $user on my steam folder grant acces to steam.

this is clearly a window side problem

**Comment by dekeonus** [comment]
I have a suspicion this is because of 

> Note that processes running under User Access Control tokens create files as the BUILTIN\Administrators SID (S-1-5-32-544), rather as a user account.

and this would mean that there is no "user" in the posix sense.
A windows process has a SID associated to it that can be a user **OR** a group (as well as a few other types, but user or group will be the most common).
A posix system has **BOTH** a user (uid) *AND* a group (gid) associated with the process.

@weskerty do you have a GroupMapping for S-1-5-32-544 (Builtin Admins Alias) **AND** S-1-5-32-545 (Builtin Users Alias) and perhaps you could also try putting S-1-5-32-544 (Builtin Admins Alias) into Mappings (i.e. user mappings) and set it to your linux user account

Ideally I'd like to set winbtrfs mount options similar to file_mode=0664 dir_mode=0775 (or file_mode=0660 dir_mode=0770) and set set_gid on the dirs. (I'm currently trying to workout if winbtrfs does allow to change the create mode used for new objects)

**Comment by marshalfightz** [comment]
Hey, 
i used this and it works well for me.

1. Create the Script 

Save the following script in a file, e.g., /usr/local/bin/update_permissions.sh:

bash

#!/bin/bash

# Script to change ownership of directories

# Ensure the script is running with sudo privileges
if [[ $EUID -ne 0 ]]; then
   exit 1
fi

# Change ownership of directories
chown -R $USER:$USER /var/mnt/games2       (mountpoint of the Steamgames)
chown -R $USER:$USER /var/mnt/games        (mountpoint of the Steamgames)

2. Make the Script Executable

Grant execute permissions to the script:

bash

sudo chmod +x /usr/local/bin/update_permissions.sh

3. Create a Systemd Unit File

Next, create a Systemd service file so that the script runs automatically at startup. Open a text editor as root to create the following file:

bash

sudo nano /etc/systemd/system/update-permissions.service

Insert this content into the file:

ini

[Unit]
Description=Update directory ownership on startup
After=network.target

[Service]
Type=oneshot
ExecStart=/usr/local/bin/update_permissions.sh
RemainAfterExit=true

[Install]
WantedBy=multi-user.target

4. Enable the Systemd Service

To ensure that the service runs automatically at startup, enable it with the following command:

bash

sudo systemctl enable update-permissions.service

5. Testing

You can manually test the service to make sure it's working:

bash

sudo systemctl start update-permissions.service

This will run the script once. Afterward, it will run automatically at the next system boot.

**Comment by marshalfightz** [comment]
Hey 

For me is working this well:

To set up a script that runs at startup on Fedora without any `echo` output, follow these steps:

### 1. Create the Script Without `echo`

Save the following script in a file, e.g., `/usr/local/bin/update_permissions.sh`:

#!/bin/bash

# Script to change ownership of directories

# Ensure the script is running with sudo privileges
if [[ $EUID -ne 0 ]]; then
   exit 1
fi

# Change ownership of directories
chown -R $USER:$USER /var/mnt/games2
chown -R $USER:$USER /var/mnt/games


### 2. Make the Script Executable

Grant execute permissions to the script:

sudo chmod +x /usr/local/bin/update_permissions.sh


### 3. Create a Systemd Unit File

Next, create a Systemd service file so that the script runs automatically at startup. Open a text editor as root to create the following file:


sudo nano /etc/systemd/system/update-permissions.service

Insert this content into the file:

[Unit]
Description=Update directory ownership on startup
After=network.target

[Service]
Type=oneshot
ExecStart=/usr/local/bin/update_permissions.sh
RemainAfterExit=true

[Install]
WantedBy=multi-user.target


### 4. Enable the Systemd Service

To ensure that the service runs automatically at startup, enable it with the following command:


sudo systemctl enable update-permissions.service

### 5. Testing

You can manually test the service to make sure it's working:


sudo systemctl start update-permissions.service

This will run the script once. Afterward, it will run automatically at the next system boot.

**Comment by marshalfightz** [comment]
Hey 

For me is working this well:

To set up a script that runs at startup on Fedora, follow these steps:

 1. Create the Script W

Save the following script in a file, e.g., `/usr/local/bin/update_permissions.sh`:

```
#!/bin/bash

# Script to change ownership of directories

# Ensure the script is running with sudo privileges
if [[ $EUID -ne 0 ]]; then
   exit 1
fi

# Change ownership of directories
chown -R $USER:$USER /var/mnt/games2
chown -R $USER:$USER /var/mnt/games


# Script to change ownership of directories

# Ensure the script is running with sudo privileges
if [[ $EUID -ne 0 ]]; then
   exit 1
fi

# Change ownership of directories
chown -R $USER:$USER /var/mnt/games2
chown -R $USER:$USER /var/mnt/games
```


2. Make the Script Executable

Grant execute permissions to the script:

`sudo chmod +x /usr/local/bin/update_permissions.sh`


3. Create a Systemd Unit File

Next, create a Systemd service file so that the script runs automatically at startup. Open a text editor as root to create the following file:


`sudo nano /etc/systemd/system/update-permissions.service`

Insert this content into the file:
```

[Unit]
Description=Update directory ownership on startup
After=network.target

[Service]
Type=oneshot
ExecStart=/usr/local/bin/update_permissions.sh
RemainAfterExit=true

[Install]
WantedBy=multi-user.target
```


4. Enable the Systemd Service

To ensure that the service runs automatically at startup, enable it with the following command:

`
sudo systemctl enable update-permissions.service`

 5. Testing

You can manually test the service to make sure it's working:

`
sudo systemctl start update-permissions.service`

This will run the script once. Afterward, it will run automatically at the next system boot.

**Comment by 42Willow** [comment]
This doesn’t really fix the root of the problem though since this issue affects non-steam related files too

**Comment by marshalfightz** [comment]
sorry there was an error i delete it

**Comment by bezo2299** [comment]
I just had a similar problem and I managed to fix it.

I'm using an external btrfs drive with steam games that were installed on linux. But when I booted into windows (WinBtrfs v1.9 installed), steam could read the games but not update them or install new games.

In the "Mappings" section on the official WinBtrfs github page it says "Processes running under User Access Control tokens create files as the BUILTIN\Administrators SID (S-1-5-32-544), rather as a user account." So I went into the registry editor to this location:
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\btrfs\Mappings
Here I created a new DWORD, named it "S-1-5-32-544" and set the value data to 1000 decimal (3e8 hex), which is the most common User ID in linux (check yours by typing "id" in linux terminal).

(I think the next step below is not necessary, but things started working, so I didn't change anything after that)
In location ...btrfs\GroupMappings there was already a DWORD with name "S-1-5-32-545". I just renamed it to "S-1-5-32-544" and left the value data unchanged.

After this everything just works. I can install/update games on linux and windows on the same drive.

**Comment by DSDV** [comment]
> In the "Mappings" section on the official WinBtrfs github page it says "Processes running under User Access Control tokens create files as the BUILTIN\Administrators SID (S-1-5-32-544), rather as a user account." So I went into the registry editor to this location: HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\btrfs\Mappings Here I created a new DWORD, named it "S-1-5-32-544" and set the value data to 1000 decimal (3e8 hex), which is the most common User ID in linux (check yours by typing "id" in linux terminal).
> 
> (I think the next step below is not necessary, but things started working, so I didn't change anything after that) In location ...btrfs\GroupMappings there was already a DWORD with name "S-1-5-32-545". I just renamed it to "S-1-5-32-544" and left the value data unchanged.

i followed those steps and after that i still get permission denied on any new file i create. 
but at least if i sudo open them i can now read them. 

ill probably start over with an ntfs drive. 
using ntfs under linux for storage seams to be better than getting "corrupted" data when using btrfs under winows :/

**Comment by bezo2299** [comment]
> i followed those steps and after that i still get permission denied on any new file i create. but at least if i sudo open them i can now read them.
> 
> ill probably start over with an ntfs drive. using ntfs under linux for storage seams to be better than getting "corrupted" data when using btrfs under winows :/
> 
> it is also the way described by valve: https://github.com/ValveSoftware/Proton/wiki/Using-a-NTFS-disk-with-Linux-and-Windows
>

Are you sure you did everything right? I used this fix a couple times and it always worked. Did you check your user ID in your linux? If it is not 1000, you need to change the D-word value from 1000 (decimal) to your user ID.

Also, maybe the second part is necessary. I left the second D-word's value unchanged because it was already set to 100, which is my group ID in linux. So check your group ID and change the D-word value in "GroupMappings" to it. Also dont forget to rename is as I described above.

I have read that NTFS can cause problems when gaming on linux. That is why I use btrfs and so far I have not had any corrupted data.

**Comment by DSDV** [comment]
Sorry for the late reply.
i think i did everything correctly. but i cant check it anymore since i got frustrated and just nuked the setup. 
my steam libary is now only for linux. 
and while i keep a spare 250GB SSD with a windows setup jic i did not boot it since (:

**Comment by Brensom** [comment]
### Cross-issue summary for dual-boot Steam users (Win ↔ Linux, Btrfs)

Linking **all open tickets** that describe the **same root cause**  
and the **same zero-re-download work-around**.

--------------------------------------------------
Root cause
--------------------------------------------------
1. Windows Steam (or any Windows process via WinBtrfs) **creates files** inside the **shared Steam library**  
   with **SID that maps to Linux UID 65534 (nobody)**.  
2. Linux Steam then **cannot overwrite / delete** those files →  
   **“Disk write failure”**, **“content file locked”**, **“Staging folder not writable”**.  
3. Games **appear fully installed** but **refuse to launch** until permissions are manually reset.

--------------------------------------------------
Tickets / issues (open)
--------------------------------------------------
- **WinBtrfs #655** ← **you are here**  
  https://github.com/maharmstone/btrfs/issues/655  
  (generic permission reset after every Windows boot)

- **Valve #12015**  
  https://github.com/ValveSoftware/steam-for-linux/issues/12015  
  (unpack fails because chunks can’t be written)

- **Valve #12139**  
  https://github.com/ValveSoftware/steam-for-linux/issues/12139  
  (moving games fails with “content file locked”)

- **Valve #11984** (UI crash-loop after permission failure)  
  https://github.com/ValveSoftware/steam-for-linux/issues/11984  

--------------------------------------------------
One-liner work-around (no re-download required)
--------------------------------------------------
```bash
# after every Windows session, before starting Steam on Linux:
sudo chown -R $USER:$USER /mnt/steam_library/steamapps
find /mnt/steam_library/steamapps -name "*.lock" -delete
rm -rf ~/.steam/steam/appcache/{httpcache,librarycache}

**Comment by Brensom** [comment]
### 🔄 Windows re-downloads **Windows builds** after Linux `chown` breaks **Linux builds**

**Short version**  
Fixing permissions with `chown -R $USER:$USER` under Linux **corrupts the Linux branch** of a game (manifest, shaders, Proton prefix, etc.).  
When I reboot into Windows, Steam **does NOT touch the Linux files**—it **sees that the Windows branch is missing** and starts downloading the **Windows build from scratch**, even though the **bulk data** (`.vpk`, `.pak`, textures, sounds) is already there.  
After the Windows download finishes, the **Linux branch is now broken again** (owner 65534), so the next Linux boot needs another `chown`, and the **infinite loop** continues.

---

### Step-by-step (same library, same AppID, **different builds**)

| Boot OS | State of `steamapps/common/Game` | Steam action | Result |
|---------|----------------------------------|--------------|--------|
| Linux   | owner `1000:1000`, **Linux build** | Launch → **OK** | — |
| Windows | owner `1000:1000` (foreign SID) | **Windows build NOT detected** | *“Install”* button appears |
| Windows | click *Install* | Steam **re-uses** existing bulk files, but **downloads Windows exe/ dll/ anticheat/ etc.** (~ few GB) |
| Linux   | owner now `65534` (nobody) | **“Missing file privileges”** | `chown -R 1000:1000` required |
| Windows | owner back to `1000` | **Windows branch invalidated again** → re-download |

---

### Key observation

The **game content** (textures, audio, levels) is **shared**; only **OS-specific binaries** and **manifest deltas** are re-fetched.  
However, the **trigger** for the re-download is **not the data itself**, but the **owner mismatch** on the **top-level folder** and the **`.acf` file**: Steam Windows **refuses to trust** any folder whose **ACL owner** is not mapped to a **known Windows SID**.  
WinBtrfs currently **cannot preserve** the original Windows SID once Linux `chown` runs, so the **Windows client marks the installation as “foreign”** and **re-downloads its own branch**.

---

### Request

Add a **per-folder“sticky-SID”** feature:

- A new **xattr** (e.g. `user.winbtrfs.sticky_sid`) that **remembers the Windows SID** that created the folder.  
- When Linux `chown` is executed, **only the POSIX uid/gid changes**; the **stored SID remains**, and WinBtrfs **continues to report that SID to Windows**.  
- Optional mount flag `sticky_sid` to enable the behaviour globally.

This would let Linux **fix its own permissions** without **invalidating the Windows branch**, breaking the **dual-boot re-download loop** once and for all.

---

### Work-arounds attempted (still loop)

- Registry `Mappings` S-1-5-32-544 → 1000, S-1-5-32-545 → 1000  
- Separate `appcache` and `.acf` copies  
- `setgid`, `umask 002`, `fastboot off`  

None prevent Windows from **re-downloading its branch** after a Linux `chown`.

---

### Environment

- WinBtrfs 1.9  
- Windows 11 23H2  
- Arch Linux, kernel 6.17.8-arch1-1, uid/gid 1000  
- Steam latest stable (both OSes)  
- Library on dedicated Btrfs partition mounted at `/mnt/steam_library`

---

### Impact

At present **a shared Steam library on Btrfs is impossible** for **dual-boot gamers** because **every OS switch** either  
- breaks **Linux launching** (permission denied) **or**  
- forces **Windows to re-fetch its build** (multi-GB download).  

A **sticky-SID** mechanism inside WinBtrfs would finally make **Btrfs the ideal cross-platform gaming filesystem**.

Thank you for considering this enhancement!

**Comment by maharmstone** [comment]
@Brensom, don't use AI to create comments, it wastes everybody's time. Future AI comments will be deleted.

If the POSIX uid of a file is the same as an entry in the Mappings table, that SID gets used regardless of what's stored in the security.NTACL xattr.
